src/server/api/controllers/orders.controller.js

```diff
@@ -0,0 +1,27 @@
+const knex = require('../../config/db');
+const moment = require('moment-timezone');
+// const HttpError = require('../lib/utils/http-error');
+
+const newOrder = async (data) => {
+  let orderId;
+  await knex('orders')
+    .insert({ user_id: data.user_id,
+            created_at: moment(Date.now()).format(),
+            status: 'NEW'})
+    .returning('id')
+    .then((id) => {
+      [orderId] = id;
+      data.items.forEach(async (item) => {
+        await knex('order_items').insert({
+          order_id: id[0],
+          product_id: item.productId,
+          quantity: item.quantity,
+        });
+      });
+    });
+  return orderId;
```

Comment:
This code mixes Promises with async/await which defeats the gain, it also misuses async/await in combination with `forEach` which will lead to unexpected behavior. 

This is fairly complicated to explain, I will try but also provide code help!

**Mixing Promises with async/await**

First of all, fundamentally, the aim of using async/await is to reach code which looks as if it is synchronous, which usually means "no indentation jumps". Compare to this:
```js
const orderId = await knex('orders')
  .insert({ 
    user_id: data.user_id,
    created_at: moment(Date.now()).format(),
    status: 'NEW',
  })
  .returning('id');

data.items.forEach(async (item) => {
  await knex('order_items').insert({
    order_id: orderId,
    product_id: item.productId,
    quantity: item.quantity,
  });
});

return orderId;
```

This code is more "flat" and gives you the impression that first we create the order, then we process the items, as two distinct synchronous steps. Of course, behind the scenes it is asynchronous thanks to Promises.

**Async/Await and array methods like forEach**

This is more or less what forEach looks like internally:
```js
function forEach(iteratorFunction) {
  const list = this;
  for (let i = 0; i < list.length; i++) {
    const currentItem = list[i];
    iteratorFunction(currentItem);
  }
}
```
Notice how there is no `await` in front of the calling of `iteratorFunction`. This means that even if you mark that function as `async` it won't be waited for.

This means that the current code will loop over the items, start of a number of asynchronous calls to the database but it won't wait for them (or any errors that happen in those calls!). This can easily lead to very bad bugs that are hard to find.
So, how can we re-write this code to act the way we intended?
The answer is a combination of `Promise.all` and `Array.map`.

We need one final piece of understanding before we look at the code. The return value of any function marked with `async` is a Promise.

That means that we can do this:
```js
async function addOrderItem(item) {
  return await knex('order_items').insert({
    order_id: orderId,
    product_id: item.productId,
    quantity: item.quantity,
  });
}
const promises = data.items.map(addOrderItem);
```
This will call `addOrderItem` for each item in the order and return a Promise.
We can then pass this array of Promises to `Promise.all` which returns a new Promise, that only resolves once **all** Promises have resolved.

```js
await Promise.all(promises);
return orderId;
```

The key takeaway here is that just because you mark a function with `async/await` does not mean whoever is calling that function will actually wait for it!

Reply:
It's worth noting that the `Promise.all` approach makes all the requests happen in parallel. 
It's also worth noting that a `Promise` is `hot` by design, this means that the moment you create a Promise is becomes in progress. 
So while the second Promise is being created inside the call to `map`, the first Promise is already in progress.

If you want to avoid this, it's better to use a plain `for...of` loop instead, to run requests serially:
```js
for (item of items) {
  await knex('order_items').insert({
    order_id: orderId,
    product_id: item.productId,
    quantity: item.quantity, 
  });
}
```

In this situation, the parallel processing is probably okay.

Another important factor is how to handle failures in this context in relation to the database. If adding an order item fails, what should happen to the overall order?
Using something like [Transactions](https://knexjs.org/#Transactions) is most likely a very good idea here.